### PR TITLE
Store links in sidebar

### PR DESCRIPTION
**Pull request by aumayr**
ref #210 

Adds a new setting to display custom links in the sidebar:

```ini
; Show custom links on top of the sidebar
sidebar-custom-links =
    Journal 2015: /journal/?time=2015
    Income Statement 2015: /income_statement/?time=2015
```

The format for the links is `[label]: [link]`. Note that `": "` can not be part of the label. 

Will be rendered like this:

screenshot

TODO:

- [ ] Add Documentation-entries.

### CODE REVIEW

**Comment by yagebu** [comment]
This seems to be something file-specific to me, so why not use `custom` directives for this?

**Comment by aumayr** [comment]
> This seems to be something file-specific to me, so why not use `custom` directives for this?

I hesitate to implement this for just this feature - it would be generally better to implement #214 (Move fava settings to source file options) and have all the settings in the beancount-file, as otherwise `fava`-specific settings are spread over two places, when they should be in just one.

What do you think about #214? Good idea? What format should we choose to store the settings in the beancount file?

**Comment by yagebu** [comment]
> I hesitate to implement this for just this feature - it would be generally better to implement #214 

I view this feature as being mostly unrelated to #214. This is not a "setting" for me, but rather something like a more flexible `query` directive. So it doesn't really make sense without the context of the beancount file (just like budgets).

> What do you think about #214? Good idea? What format should we choose to store the settings in the beancount file?

Probably a good idea, although not a pressing issue IMHO. I guess the `custom` directives are the best choice right now. Before we tackle this, we should evaluate if we really need all the settings we currently have.

**Comment by aumayr** [comment]
> > I hesitate to implement this for just this feature - it would be generally better to implement #214 
> 
I view this feature as being mostly unrelated to #214. This is not a "setting" for me, but rather something like a more flexible `query` directive. So it > doesn't really make sense without the context of the beancount file (just like budgets).

Implemented in 9f7cfb5

> > What do you think about #214? Good idea? What format should we choose to store the settings in the beancount file?
> 
Probably a good idea, although not a pressing issue IMHO. I guess the `custom` directives are the best choice right now. Before we tackle this, we should > evaluate if we really need all the settings we currently have. 

Let's discuss this in #214. Can you post your thoughts on the various settings there?

**Comment by aumayr** [comment]
This works for me. The custom links can be specified with

```
2016-04-01 custom "fava-sidebar-link" "Income 2014" "/income_statement?time=2014"
```

Only a help entry is missing.

**Comment by yagebu** [comment]
Does this still work after #259? If so, can you rebase?

I guess we'll have to wait for a new beancount release that contains the custom directive before we can merge this.

File: fava/api/__init__.py [hunk 1]
```diff
@@ -626,3 +626,10 @@ def account_open_metadata(self, account_name):
             if isinstance(posting, Open):
                 return posting.meta
         return {}
+
+    def sidebar_links(self):
+        # 2016-04-01 custom "fava-sidebar-link" "Income 2014" "/income_statement?time=2014"  # noqa
+        return [(entry.values[0].value, entry.values[1].value)
```
Comment (yagebu) [inline fava/api/__init__.py]: This runs through all entries once on each page load which is horribly inefficient for such a small feature.

Comment (aumayr) [inline fava/api/__init__.py]: Good point! Fixed in 3c61688

**Comment by aumayr** [comment]
> Does this still work after #259? If so, can you rebase?

It works, and I rebased.

> I guess we'll have to wait for a new beancount release that contains the custom directive before we can merge this.

Agreed. But I think @blais would not hesitate to release a new version with the custom directives if requested.

**Comment by aumayr** [comment]
Tests keep failing because the `custom` directive is not part of the last official PyPI release of beancount. 

@blais Would you consider releasing a new version to PyPI with the `custom` directive included?

**Comment by yagebu** [comment]
> It works, and I rebased.

I think something went wrong with the rebase, as it shows 40 commits here. But the diff is not big, so it's probably easiest to just create one new commit with all the changes:

    git merge master       # add new commits since the rebase
    git reset master
    # then git add and git commit

**Comment by blais** [comment]
Done.
https://pypi.python.org/pypi/beancount/
2.0b10

**Comment by aumayr** [comment]
@blais The files on PyPI seem to be missing, so 2.0b10 cannot be found:

https://pypi.python.org/pypi/beancount/json

**Comment by blais** [comment]
PyPI is pissing me off.
Tested manually, now I can't upload to the same filename.
I baked another version b11.
Try onw.

**Comment by aumayr** [comment]
It works now, thanks @blais!

> PyPI is pissing me off.

Sorry for causing trouble!

**Comment by aumayr** [comment]
@yagebu I, again, messed something up with git. I will add the missing documentation here, and create a new PR once I am finished.

**Comment by aumayr** [comment]
Because the git history was messed up, this PR will be closed, and the successor can be found here:

#279

**Pull request closed without merging**